File: harness/bootstrap_grant_issuance_contract/documents.py
"""Signed Policy, Request, and durable Receipt structural validation."""

from __future__ import annotations

import hashlib
import re
from typing import Any

from .canonical import ContractError, bounded_canonical_json, self_digest
from .constants import (CANONICALIZATION, MAX_POLICY_BYTES, MAX_POLICY_VALIDITY_MS,
                        MAX_RECEIPT_BYTES, MAX_REQUEST_BYTES, MAX_REQUEST_VALIDITY_MS,
                        MAX_TTL_MS, POLICY_API, POLICY_DOMAIN, RECEIPT_API, RECEIPT_DOMAIN,
                        RECORD_KEY_DOMAIN, REQUEST_API, REQUEST_DOMAIN)
from .shape import (budget_covers, enum, integer, require_keys, sha256, text,
                    validate_budget, validate_capability, validate_principal,
                    validate_profile_id, validate_request_bindings, validate_scope,
                    validate_signature, validate_task_binding)

POLICY_FIELDS = {
    "api_version", "budget", "canonicalization", "capability", "disposition",
    "effect_id", "kind", "max_ttl_ms", "policy_id", "policy_sha256", "profile_id",
    "scope", "signature", "subject", "task_binding", "trust_epoch",
    "trust_root_sha256", "validity",
}
REQUEST_FIELDS = {
    "api_version", "bindings", "budget", "canonicalization", "capability", "effect_id",
    "expires_at_unix_ms", "idempotency_key", "kind", "policy_sha256", "profile_id",
    "request_sha256", "requested_at_unix_ms", "requested_ttl_ms", "scope", "signature",
    "subject", "task_binding", "trust_epoch", "trust_root_sha256",
}
RECEIPT_FIELDS = {
    "api_version", "canonicalization", "decision", "denial_reason",
    "grant_envelope_sha256", "grant_id", "grant_sha256", "kind", "ledger_sequence",
    "policy_sha256", "prior_receipt_sha256", "profile_id", "receipt_sha256",
    "record_key_sha256", "request_sha256", "signature", "stored_at_unix_ms",
    "trust_epoch", "trust_root_sha256",
}


def policy_sha256(value: dict[str, Any]) -> str:
    return self_digest(POLICY_DOMAIN, value, "policy_sha256", MAX_POLICY_BYTES,
                       "BootstrapGrantPolicy", signed=True)


def validate_policy(value: Any, profile_hash: str) -> dict[str, Any]:
    node = require_keys(value, "BootstrapGrantPolicy", POLICY_FIELDS)
    bounded_canonical_json(node, MAX_POLICY_BYTES, "BootstrapGrantPolicy")
    _require_envelope(node, POLICY_API, "BootstrapGrantPolicy")
    validate_profile_id(node["profile_id"], "policy.profile_id")
    text(node["policy_id"], "policy.policy_id")
    enum(node["disposition"], "policy.disposition", ("allow", "deny"))
    if node["effect_id"] != "repo.read":
        raise ContractError("policy.effect_id must be repo.read")
    validate_capability(node["capability"], "policy.capability")
    validate_principal(node["subject"], "policy.subject")
    validate_task_binding(node["task_binding"], "policy.task_binding")
    validate_scope(node["scope"], "policy.scope")
    validate_budget(node["budget"], "policy.budget")
    integer(node["max_ttl_ms"], "policy.max_ttl_ms", 1, MAX_TTL_MS)
    _validate_policy_validity(node["validity"])
    _validate_common_authority(node, "policy")
    validate_signature(node["signature"], "policy.signature", profile_hash)
    _require_self_digest(node, "policy_sha256", policy_sha256, "policy")
    return node


def _validate_policy_validity(value: Any) -> None:
    keys = {"expires_at_unix_ms", "not_before_unix_ms"}
    node = require_keys(value, "policy.validity", keys)
    start = integer(node["not_before_unix_ms"], "policy.validity.not_before_unix_ms",
                    0, 2**63 - 1)
    end = integer(node["expires_at_unix_ms"], "policy.validity.expires_at_unix_ms",
                  0, 2**63 - 1)
    if not start < end or end - start > MAX_POLICY_VALIDITY_MS:
        raise ContractError("policy validity must be ordered within 24 hours")


def request_sha256(value: dict[str, Any]) -> str:
    return self_digest(REQUEST_DOMAIN, value, "request_sha256", MAX_REQUEST_BYTES,
                       "BootstrapGrantRequest", signed=True)


def validate_request(value: Any, profile_hash: str) -> dict[str, Any]:
    node = require_keys(value, "BootstrapGrantRequest", REQUEST_FIELDS)
    bounded_canonical_json(node, MAX_REQUEST_BYTES, "BootstrapGrantRequest")
    _require_envelope(node, REQUEST_API, "BootstrapGrantRequest")
    validate_profile_id(node["profile_id"], "request.profile_id")
    if node["effect_id"] != "repo.read":
        raise ContractError("request.effect_id must be repo.read")
    validate_request_bindings(node["bindings"], "request.bindings")
    validate_capability(node["capability"], "request.capability")
    validate_principal(node["subject"], "request.subject")
    validate_task_binding(node["task_binding"], "request.task_binding")
    validate_scope(node["scope"], "request.scope")
    validate_budget(node["budget"], "request.budget")
    _validate_request_time_and_key(node)
    _validate_common_authority(node, "request")
    validate_signature(node["signature"], "request.signature", profile_hash)
    sha256(node["policy_sha256"], "request.policy_sha256")
    _require_self_digest(node, "request_sha256", request_sha256, "request")
    return node


def _validate_request_time_and_key(node: dict[str, Any]) -> None:
    key = node["idempotency_key"]
    if not isinstance(key, str) or re.fullmatch(r"[A-Za-z0-9._:@+\-]{16,128}", key) is None:
        raise ContractError("request.idempotency_key must be 16..128 closed visible ASCII bytes")
    start = integer(node["requested_at_unix_ms"], "request.requested_at_unix_ms",
                    0, 2**63 - 1)
    end = integer(node["expires_at_unix_ms"], "request.expires_at_unix_ms", 0, 2**63 - 1)
    if not start < end or end - start > MAX_REQUEST_VALIDITY_MS:
        raise ContractError("request validity must be ordered within five minutes")
    integer(node["requested_ttl_ms"], "request.requested_ttl_ms", 1, MAX_TTL_MS)


def validate_policy_request(policy: dict[str, Any], request: dict[str, Any]) -> None:
    equal_fields = ("capability", "effect_id", "profile_id", "scope", "subject",
                    "task_binding", "trust_epoch", "trust_root_sha256")
    if any(request[field] != policy[field] for field in equal_fields):
        raise ContractError("Policy and Request exact fields differ")
    if request["policy_sha256"] != policy["policy_sha256"]:
        raise ContractError("Request does not bind the exact Policy digest")
    if not budget_covers(policy["budget"], request["budget"]):
        raise ContractError("Request budget exceeds Policy budget")
    if request["requested_ttl_ms"] > policy["max_ttl_ms"]:
        raise ContractError("Request TTL exceeds Policy maximum")
    validity = policy["validity"]
    if not (validity["not_before_unix_ms"] <= request["requested_at_unix_ms"] <
            request["expires_at_unix_ms"] <= validity["expires_at_unix_ms"]):
        raise ContractError("Request validity is outside Policy validity")


def record_key_sha256(idempotency_key: str) -> str:
    if (not isinstance(idempotency_key, str) or
            re.fullmatch(r"[A-Za-z0-9._:@+\-]{16,128}", idempotency_key) is None):
        raise ContractError("idempotency key must be 16..128 closed visible ASCII bytes")
    return hashlib.sha256(RECORD_KEY_DOMAIN + idempotency_key.encode("ascii")).hexdigest()


def receipt_sha256(value: dict[str, Any]) -> str:
    return self_digest(RECEIPT_DOMAIN, value, "receipt_sha256", MAX_RECEIPT_BYTES,
                       "GrantIssuanceReceipt", signed=True)


def validate_receipt(value: Any, profile_hash: str) -> dict[str, Any]:
    node = require_keys(value, "GrantIssuanceReceipt", RECEIPT_FIELDS)
    bounded_canonical_json(node, MAX_RECEIPT_BYTES, "GrantIssuanceReceipt")
    _require_envelope(node, RECEIPT_API, "GrantIssuanceReceipt")
    validate_profile_id(node["profile_id"], "receipt.profile_id")
    enum(node["decision"], "receipt.decision", ("denied", "issued"))
    integer(node["ledger_sequence"], "receipt.ledger_sequence", 1, 2**63 - 1)
    integer(node["stored_at_unix_ms"], "receipt.stored_at_unix_ms", 0, 2**63 - 1)
    for field in ("policy_sha256", "receipt_sha256", "record_key_sha256",
                  "request_sha256", "trust_root_sha256"):
        sha256(node[field], f"receipt.{field}")
    for field in ("grant_envelope_sha256", "grant_sha256", "prior_receipt_sha256"):
        if node[field] is not None:
            sha256(node[field], f"receipt.{field}")
    validate_signature(node["signature"], "receipt.signature", profile_hash)
    _validate_receipt_grant_fields(node)
    _validate_common_authority(node, "receipt")
    _require_self_digest(node, "receipt_sha256", receipt_sha256, "receipt")
    return node


def _validate_receipt_grant_fields(node: dict[str, Any]) -> None:
    fields = (node["grant_id"], node["grant_sha256"], node["grant_envelope_sha256"])
    if node["decision"] == "issued":
        if node["denial_reason"] is not None or any(value is None for value in fields):
            raise ContractError("issued receipt requires all Grant identities and no denial")
        expected_id = f"capability-grant-{node['grant_sha256']}"
        if node["grant_id"] != expected_id:
            raise ContractError("receipt grant_id does not match grant_sha256")
    elif node["denial_reason"] != "policy_denied" or any(value is not None for value in fields):
        raise ContractError("denied receipt must have policy_denied and null Grant identities")


def validate_receipt_relations(policy: dict[str, Any], request: dict[str, Any],
                               receipt: dict[str, Any]) -> None:
    expected_decision = "issued" if policy["disposition"] == "allow" else "denied"
    if receipt["decision"] != expected_decision:
        raise ContractError("receipt decision differs from authenticated Policy disposition")
    relations = (("policy_sha256", policy["policy_sha256"]),
                 ("request_sha256", request["request_sha256"]),
                 ("record_key_sha256", record_key_sha256(request["idempotency_key"])),
                 ("trust_root_sha256", request["trust_root_sha256"]),
                 ("trust_epoch", request["trust_epoch"]),
                 ("profile_id", request["profile_id"]))
    if any(receipt[field] != expected for field, expected in relations):
        raise ContractError("receipt does not bind Policy, Request, record key, or root")
    stored = receipt["stored_at_unix_ms"]
    if not request["requested_at_unix_ms"] <= stored < request["expires_at_unix_ms"]:
        raise ContractError("receipt stored time is outside Request validity")


def _validate_common_authority(node: dict[str, Any], label: str) -> None:
    sha256(node["trust_root_sha256"], f"{label}.trust_root_sha256")
    integer(node["trust_epoch"], f"{label}.trust_epoch", 1, 2**63 - 1)


def _require_envelope(node: dict[str, Any], api: str, kind: str) -> None:
    if (node["api_version"], node["canonicalization"], node["kind"]) != (
            api, CANONICALIZATION, kind):
        raise ContractError(f"{kind} envelope drifted from v1")


def _require_self_digest(node: dict[str, Any], field: str, function: Any, label: str) -> None:
    sha256(node[field], f"{label}.{field}")
    if node[field] != function(node):
        raise ContractError(f"{label} self digest does not match")

File: forge-core/internal/bootstrapgrantauthority/documents.go
package bootstrapgrantauthority

import (
	"crypto/sha256"
	"fmt"
)

var policyKeys = []string{
	"api_version", "budget", "canonicalization", "capability", "disposition", "effect_id",
	"kind", "max_ttl_ms", "policy_id", "policy_sha256", "profile_id", "scope", "signature",
	"subject", "task_binding", "trust_epoch", "trust_root_sha256", "validity",
}

var requestKeys = []string{
	"api_version", "bindings", "budget", "canonicalization", "capability", "effect_id",
	"expires_at_unix_ms", "idempotency_key", "kind", "policy_sha256", "profile_id",
	"request_sha256", "requested_at_unix_ms", "requested_ttl_ms", "scope", "signature",
	"subject", "task_binding", "trust_epoch", "trust_root_sha256",
}

// Policy is a structurally valid and cryptographically authenticated Policy.
type Policy struct{ document map[string]any }

// Request is a structurally valid and cryptographically authenticated request.
type Request struct{ document map[string]any }

// DecodePolicy authenticates a strict canonical Policy against the pinned root.
func DecodePolicy(data []byte, trust *Trust) (*Policy, error) {
	if trust == nil {
		return nil, fmt.Errorf("Trust is required")
	}
	document, err := decodeCanonical(data, maxPolicyBytes)
	if err != nil {
		return nil, err
	}
	if err = validatePolicy(document, trust); err != nil {
		return nil, err
	}
	return &Policy{document: document}, nil
}

// DecodeRequest authenticates a strict request and its exact Policy relations.
func DecodeRequest(data []byte, trust *Trust, policy *Policy) (*Request, error) {
	if trust == nil || policy == nil {
		return nil, fmt.Errorf("Trust and Policy are required")
	}
	document, err := decodeCanonical(data, maxRequestBytes)
	if err != nil {
		return nil, err
	}
	if err = validateRequest(document, trust); err != nil {
		return nil, err
	}
	if err = validatePolicyRequest(policy.document, document); err != nil {
		return nil, err
	}
	return &Request{document: document}, nil
}

func validatePolicy(document map[string]any, trust *Trust) error {
	if err := requireKeys(document, policyKeys...); err != nil {
		return fmt.Errorf("BootstrapGrantPolicy: %w", err)
	}
	if err := validatePolicyShape(document); err != nil {
		return err
	}
	if err := validateAuthorityBinding(document, trust, "Policy"); err != nil {
		return err
	}
	return validateSignedDocument(document, "policy_sha256", policyDomain,
		policySignatureDomain, maxPolicyBytes, "Policy", trust, "policy_sign")
}

func validatePolicyShape(document map[string]any) error {
	if err := validateDocumentEnvelope(document, policyAPI, "BootstrapGrantPolicy"); err != nil {
		return err
	}
	if err := validateTextField(document, "policy_id", 160); err != nil {
		return err
	}
	disposition, err := stringValue(document, "disposition")
	if err != nil || !oneOf(disposition, "allow", "deny") || document["effect_id"] != "repo.read" {
		return fmt.Errorf("Policy disposition or effect is invalid")
	}
	if _, err = validateCapabilityNode(document["capability"], "Policy capability"); err != nil {
		return err
	}
	if _, err = validatePrincipalNode(document["subject"], "Policy subject"); err != nil {
		return err
	}
	if _, err = validateTaskNode(document["task_binding"], "Policy task_binding"); err != nil {
		return err
	}
	if _, err = validateScopeNode(document["scope"], "Policy scope"); err != nil {
		return err
	}
	if _, err = validateBudgetNode(document["budget"], "Policy budget"); err != nil {
		return err
	}
	return validatePolicyWindow(document)
}

func validatePolicyWindow(document map[string]any) error {
	maximum, err := intValue(document, "max_ttl_ms")
	if err != nil || validateRange(maximum, "Policy max_ttl_ms", 1, maxTTLMillis) != nil {
		return fmt.Errorf("Policy max_ttl_ms is invalid")
	}
	validity, err := objectValue(document, "validity")
	if err != nil || requireKeys(validity, "expires_at_unix_ms", "not_before_unix_ms") != nil {
		return fmt.Errorf("Policy validity fields are invalid")
	}
	start, startErr := intValue(validity, "not_before_unix_ms")
	end, endErr := intValue(validity, "expires_at_unix_ms")
	if startErr != nil || endErr != nil || start < 0 || end <= start || end-start > maxPolicyWindow {
		return fmt.Errorf("Policy validity must be ordered within 24 hours")
	}
	return nil
}

func validateRequest(document map[string]any, trust *Trust) error {
	if err := requireKeys(document, requestKeys...); err != nil {
		return fmt.Errorf("BootstrapGrantRequest: %w", err)
	}
	if err := validateRequestShape(document); err != nil {
		return err
	}
	if err := validateAuthorityBinding(document, trust, "Request"); err != nil {
		return err
	}
	if equal, _ := sameCanonical(document["subject"], trust.keys["request_auth"].principal); !equal {
		return fmt.Errorf("request-auth principal must equal Request subject")
	}
	return validateSignedDocument(document, "request_sha256", requestDomain,
		requestSignatureDomain, maxRequestBytes, "Request", trust, "request_auth")
}

func validateRequestShape(document map[string]any) error {
	if err := validateDocumentEnvelope(document, requestAPI, "BootstrapGrantRequest"); err != nil {
		return err
	}
	if document["effect_id"] != "repo.read" {
		return fmt.Errorf("Request effect must be repo.read")
	}
	validators := []func() error{
		func() error {
			_, err := validateCapabilityNode(document["capability"], "Request capability")
			return err
		},
		func() error { _, err := validatePrincipalNode(document["subject"], "Request subject"); return err },
		func() error { _, err := validateTaskNode(document["task_binding"], "Request task_binding"); return err },
		func() error { _, err := validateScopeNode(document["scope"], "Request scope"); return err },
		func() error { _, err := validateBudgetNode(document["budget"], "Request budget"); return err },
		func() error { return validateRequestBindings(document["bindings"]) },
		func() error { return validateRequestWindow(document) },
	}
	for _, validator := range validators {
		if err := validator(); err != nil {
			return err
		}
	}
	return validateHashField(document, "policy_sha256", "Request policy_sha256")
}

func validateRequestBindings(value any) error {
	node, ok := value.(map[string]any)
	if !ok || requireKeys(node, "context_sha256", "source_revision", "source_tree_sha256") != nil {
		return fmt.Errorf("Request bindings fields are invalid")
	}
	if err := validateHashField(node, "context_sha256", "Request context_sha256"); err != nil {
		return err
	}
	if err := validateHashField(node, "source_tree_sha256", "Request source_tree_sha256"); err != nil {
		return err
	}
	return validateTextField(node, "source_revision", 160)
}

func validateRequestWindow(document map[string]any) error {
	key, err := stringValue(document, "idempotency_key")
	if err != nil || !idempotencyPattern.MatchString(key) {
		return fmt.Errorf("Request idempotency_key is invalid")
	}
	start, startErr := intValue(document, "requested_at_unix_ms")
	end, endErr := intValue(document, "expires_at_unix_ms")
	ttl, ttlErr := intValue(document, "requested_ttl_ms")
	if startErr != nil || endErr != nil || start < 0 || end <= start || end-start > maxRequestAge {
		return fmt.Errorf("Request freshness window is invalid")
	}
	if ttlErr != nil || validateRange(ttl, "requested_ttl_ms", 1, maxTTLMillis) != nil {
		return fmt.Errorf("Request TTL is invalid")
	}
	return nil
}

func validatePolicyRequest(policy, request map[string]any) error {
	fields := []string{"capability", "effect_id", "profile_id", "scope", "subject",
		"task_binding", "trust_epoch", "trust_root_sha256"}
	for _, field := range fields {
		equal, err := sameCanonical(policy[field], request[field])
		if err != nil || !equal {
			return fmt.Errorf("Policy and Request exact field %s differs", field)
		}
	}
	if request["policy_sha256"] != policy["policy_sha256"] {
		return fmt.Errorf("Request does not bind the exact Policy")
	}
	policyBudget, _ := objectValue(policy, "budget")
	requestBudget, _ := objectValue(request, "budget")
	if !budgetCovers(policyBudget, requestBudget) {
		return fmt.Errorf("Request budget exceeds Policy budget")
	}
	return validatePolicyRequestTime(policy, request)
}

func validatePolicyRequestTime(policy, request map[string]any) error {
	policyTTL, _ := intValue(policy, "max_ttl_ms")
	requestTTL, _ := intValue(request, "requested_ttl_ms")
	requested, _ := intValue(request, "requested_at_unix_ms")
	expires, _ := intValue(request, "expires_at_unix_ms")
	validity, _ := objectValue(policy, "validity")
	notBefore, _ := intValue(validity, "not_before_unix_ms")
	policyExpires, _ := intValue(validity, "expires_at_unix_ms")
	if requestTTL > policyTTL || requested < notBefore || expires > policyExpires {
		return fmt.Errorf("Request TTL or freshness exceeds Policy")
	}
	return nil
}

func validateAuthorityBinding(document map[string]any, trust *Trust, label string) error {
	if document["profile_id"] != contractProfileID || document["trust_root_sha256"] != trust.rootHash ||
		document["trust_epoch"] != trust.epoch {
		return fmt.Errorf("%s authority binding is invalid", label)
	}
	return nil
}

func validateSignedDocument(document map[string]any, digestField string, digestDomain,
	signatureDomain []byte, maximum int, label string, trust *Trust, usage string) error {
	claimed, err := stringValue(document, digestField)
	if err != nil || validateHash(claimed, label+" digest") != nil {
		return fmt.Errorf("%s self digest is invalid", label)
	}
	computed, err := selfDigest(digestDomain, document, digestField, maximum, label, true)
	if err != nil || computed != claimed {
		return fmt.Errorf("%s self digest does not match", label)
	}
	signature, err := validateSignatureNode(document["signature"], label, trust.profileHash)
	if err != nil || signature["key_id"] != trust.keys[usage].id {
		return fmt.Errorf("%s signature key binding is invalid", label)
	}
	return verifyDigest(trust.keys[usage].publicKey, signatureDomain, claimed,
		signature["signature_base64url"].(string))
}

func validateDocumentEnvelope(document map[string]any, api, kind string) error {
	for key, expected := range map[string]string{"api_version": api,
		"canonicalization": canonicalization, "kind": kind, "profile_id": contractProfileID} {
		if err := requireLiteral(document, key, expected); err != nil {
			return err
		}
	}
	return nil
}

func validateHashField(document map[string]any, field, label string) error {
	value, err := stringValue(document, field)
	if err != nil {
		return err
	}
	return validateHash(value, label)
}

func recordKey(idempotencyKey string) string {
	hash := sha256.New()
	_, _ = hash.Write(recordKeyDomain)
	_, _ = hash.Write([]byte(idempotencyKey))
	return fmt.Sprintf("%x", hash.Sum(nil))
}
